design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,137 @@
+    1. The operator uses `objectBucketClaim.spec.secretName` to get the S3 access keys secret.  
+    1. The operator uses `objectBucketClaim.spec.storageClassName` to get the `Service` endpoint of the object store.
+1. The operator uses the object store endpoint and access keys for an S3 "make bucket" call.
+1. The operator creates a ConfigMap in the namespace of the `ObjectBucketClaim` with relevant connection data of the bucket.
+1. An app Pod may then mount the Secret and the ConfigMap to begin accessing the bucket. 
+
+**Use Case:** Delete an Object Bucket
+
+_As a Kubernetes user, I want to delete `ObjectBucketClaim` instances and cleanup generated API resources._
+
+1. The user deletes the `ObjectBucketClaim` via `kubectl delete ...`.
+1. The `ObjectBucketClaim` is marked for deletion and left in the foreground.
+1. The respective ConfigMap is deleted.
+1. The `ObjectBucketClaim` is garbage collected.
+
+---
+
+## API Specifications
+
+#### ObjectBucketClaim CRD
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: objectbucketclaims.rook.io
+spec:
+  group: rook.io
+  version: v1alpha2
+  scope: namespaced
+  names:
+      kind: ObjectBucketClaim
+      listKind: ObjectBucketClaimList
+      singular: objectbucketclaim
+      plural: objectbucketclaims
+```
+
+#### ObjectBucketClaim
+
+```yaml
+apiVersion: rook.io/v1alpha2
+kind: ObjectBucketClaim
+metadata:
+  name: my-bucket-1
+  namespace: dev-user
+  labels:
+    rook.io/bucket-provisioner:
+    rook.io/object-bucket-claim: my-bucket-1 [1]
+spec:
+  storageClassName: some-object-store
+  secretName: my-s3-key-pair
+  generateBucketName: prefix [2]
+```
+
+1. Added by the rook operator.
+1. As with `metadata.generateName`, the operator will append a hyphen followed by random characters to the string given here.
+
+#### Access Keys Secret  
+  
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-s3-key-pair
+  namespace: dev-user
+data:
+  accessKeyId: <base64 encoded string>
+  secretAccessKey: <base64 encoded string>
+```
+
+#### ObjectBucketClaim ConfigMap
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: rook-object-bucket-my-bucket-1
+  namespace: dev-user
+  labels:
+    rook.io/object-bucket-claim-controller:
+    rook.io/object-bucket-claim: my-bucket-1
+  ownerReferences:  [1]
+  - name: my-bucket-1
+    uid: 1234-qwer-4321-rewq
+    apiVersion: rook.io/v1alpha2
+    kind: `ObjectBucketClaim`
+    blockOwnerDeletion: true 
+data:
+  ROOK_BUCKET_HOST: http://my-store-url-or-ip
+  ROOK_BUCKET_NAME: my-bucket-1
+  ROOK_BUCKET_PORT: 80
+  ROOK_BUCKET_SSL: no
+```
+
+1. Treat the configMap as a child of the `ObjectBucketClaim` for garbage collection.
+
+#### ObjectStore StorageClass
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: some-object-store
+provisioner: rook.io/object-bucket-claim-controller
+parameters:
+  objectStoreService: my-store
```

Comment:
In EdgeFS serviceName "my-store" will be exposed as generic Kubernetes service. Having service name defined would help EdgeFS operator (or maybe even be done in a generic way also) to determine URL endpoint to call. The bucket then can be created via general API call which I hope can be part of Rook framework itself. Essentially API has to mimic "curl http://URL:PORT/bucketName -X PUT" semantic, handle authorization and errors.

Reply:
I think I'm misunderstanding the question.  Looking at minio and edgefs: they both generate Service API objects.   Referencing the Service in the StorageClass enables the operator to `Get()` the Service, from which it will get the IP and Port that the S3 app is listening on.  In this way, admins can "publish" one or more object stores and the user is not required to know anything about them.